File: go/svg_renderer/svg_renderer.go
// SPDX-License-Identifier: MIT
// Copyright © 2023  Bilal Djelassi

package svg_renderer

import (
	"io"
	"fmt"
	"log"
	pt "klotski-puzzle/puzzle_types"
)

type CustomRenderGridCallable func (
	renderer *SvgRenderer,
	svg *XmlWriter,
	grid *pt.Grid,
	title string)

type CustomRenderPieceCallable func (
	renderer *SvgRenderer,
	svg *XmlWriter,
	grid *pt.Piece,
	step *pt.Step)

type SvgRenderer struct {
	cellGap float64
	gridGap float64
	unitSize float64
	titleHeight float64
	offsetX float64
	offsetY float64
	borderRadius float64
	gridsPerRow int
	gridSizeX int
	gridSizeY int
	includeBackground bool
	properties map[string]string

	CustomRenderGrid CustomRenderGridCallable
	CustomRenderPiece CustomRenderPieceCallable
}

func NewSvgRenderer(grid *pt.Grid, theme string) *SvgRenderer {
	return &SvgRenderer{
		cellGap: 4.0,
		gridGap: 16.0,
		unitSize: 20.0,
		titleHeight: 12.0,
		offsetX: 0.5,
		offsetY: 0.5,
		borderRadius: 2.0,
		gridsPerRow: 10,
		gridSizeX: grid.SizeX,
		gridSizeY: grid.SizeY,
		includeBackground: false,
		properties: GetRendererTheme(theme),
	}
}

func GetRendererTheme(name string) map[string]string {
	simpleTheme := map[string]string{
		"fontFamily": "'Consolas','PT Mono','DejaVu Serif Mono',monospace",
		"fontSize": "16",
		"strokeWidthGrid": "1",
		"strokeWidthPiece": "1",
		"backgroundColor": "#ffffff",
		"textColor": "#0d1e26",
		"arrowColor": "#ffffff",
		"fillColorGrid": "#ecf4f9",
		"strokeColorGrid": "#8f9ca3",
		"fillColorHiPieceA": "#00aaff",
		"fillColorLoPieceA": "#8f9ca3",
		"fillColorHiPieceB": "#00aaff",
		"fillColorLoPieceB": "#8f9ca3",
		"fillColorHiPieceC": "#00aaff",
		"fillColorLoPieceC": "#8f9ca3",
		"fillColorHiPieceD": "#00aaff",
		"fillColorLoPieceD": "#8f9ca3",
		"strokeColorHiPieceA": "#0d1e26",
		"strokeColorLoPieceA": "#0d1e26",
		"strokeColorHiPieceB": "#0d1e26",
		"strokeColorLoPieceB": "#0d1e26",
		"strokeColorHiPieceC": "#0d1e26",
		"strokeColorLoPieceC": "#0d1e26",
		"strokeColorHiPieceD": "#0d1e26",
		"strokeColorLoPieceD": "#0d1e26",
	}

	colorfulTheme := map[string]string{
		"fontFamily": "'Consolas','PT Mono','DejaVu Serif Mono',monospace",
		"fontSize": "16",
		"strokeWidthGrid": "1",
		"strokeWidthPiece": "1",
		"backgroundColor": "#ffffff",
		"textColor": "#0d1e26",
		"arrowColor": "#ffffff",
		"fillColorGrid": "#ecf4f9",
		"strokeColorGrid": "#8f9ca3",
		"fillColorHiPieceA": "#ec433b",
		"fillColorLoPieceA": "#ec433b",
		"fillColorHiPieceB": "#93a707",
		"fillColorLoPieceB": "#93a707",
		"fillColorHiPieceC": "#3b98e0",
		"fillColorLoPieceC": "#3b98e0",
		"fillColorHiPieceD": "#e5b01a",
		"fillColorLoPieceD": "#e5b01a",
		"strokeColorHiPieceA": "#dc322f",
		"strokeColorLoPieceA": "#dc322f",
		"strokeColorHiPieceB": "#859900",
		"strokeColorLoPieceB": "#859900",
		"strokeColorHiPieceC": "#268bd2",
		"strokeColorLoPieceC": "#268bd2",
		"strokeColorHiPieceD": "#dba700",
		"strokeColorLoPieceD": "#dba700",
	}

	themes := map[string]map[string]string{
		"simple": simpleTheme,
		"colorful": colorfulTheme,
	}
	return themes[name]
}

func (renderer *SvgRenderer) Property(name string) string {
	value, found := renderer.properties[name]
	if !found {
		log.Fatalf("cannot find property: %s\n", name)
	}
	return value
}

func (renderer *SvgRenderer) getGridsOnXYAxis(gridCount int) (gridsOnXAxis, gridsOnYAxis int) {
	if gridCount <= renderer.gridsPerRow {
		gridsOnXAxis = gridCount
		gridsOnYAxis = 1
	} else {
		gridsOnXAxis = renderer.gridsPerRow
		gridsOnYAxis = 1 + (gridCount - 1) / renderer.gridsPerRow
	}
	return
}

func (renderer *SvgRenderer) getGridSize(includeTitle bool) (gridSizeX, gridSizeY float64) {
	gridSizeX = renderer.cellGap +
		float64(renderer.gridSizeX) * (renderer.unitSize + renderer.cellGap)

    gridSizeY = renderer.cellGap +
		float64(renderer.gridSizeY) * (renderer.unitSize + renderer.cellGap)

	if (includeTitle) {
		gridSizeY += renderer.titleHeight
	}
	return
}

func (renderer *SvgRenderer) getSvgSize(gridCount int, includeTitle bool) (svgSizeX, svgSizeY float64) {
	gridsOnXAxis, gridsOnYAxis := renderer.getGridsOnXYAxis(gridCount)
	gridSizeX, gridSizeY := renderer.getGridSize(includeTitle)

	svgSizeX = renderer.gridGap + float64(gridsOnXAxis) * (gridSizeX + renderer.gridGap)
	svgSizeY = renderer.gridGap + float64(gridsOnYAxis) * (gridSizeY + renderer.gridGap)
	return
}

func (renderer *SvgRenderer) gridPositionToXYCoord(x int, y int, includeTitle bool) (gridX, gridY float64) {
	gridSizeX, gridSizeY := renderer.getGridSize(includeTitle)

	gridX = renderer.offsetX + renderer.gridGap +
		float64(x) * (gridSizeX + renderer.gridGap)

    gridY = renderer.offsetY + renderer.gridGap +
		float64(y) * (gridSizeY + renderer.gridGap)
	
	if (includeTitle) {
		gridY += renderer.titleHeight
	}
	return
}

func (renderer *SvgRenderer) piecePositionToXYCoord(x, y int) (pieceX, pieceY float64) {
	pieceX = renderer.cellGap + float64(x) * (renderer.unitSize + renderer.cellGap)
	pieceY = renderer.cellGap + float64(y) * (renderer.unitSize + renderer.cellGap)
	return
}

func (renderer *SvgRenderer) nextGridPosition(x, y int) (int, int) {
	if x+1 >= renderer.gridsPerRow {
		return 0, y+1
	}
	return x+1, y
}

func (renderer *SvgRenderer) RenderGrid(out io.Writer, grid *pt.Grid) {
	svgSizeX, svgSizeY := renderer.getSvgSize(1, false)

	svg := NewXmlWriter(out)
	svg.
		Decl().
		Root("svg").
		Attr("width", svgSizeX).
		Attr("height", svgSizeY).
		Attr("xmlns", "http://www.w3.org/2000/svg")

	if renderer.includeBackground {
		svg.
			Elem("rect").
			Attr("width", "100%").
			Attr("height", "100%").
			Attr("fill", renderer.Property("backgroundColor")).
			Term()
	}

	renderer.render(svg, 0, 0, grid, nil, "")
	svg.Term()
}

func (renderer *SvgRenderer) RenderGrids(out io.Writer, grid *pt.Grid, moves []pt.Move) {
	gridCount := 1 + len(moves) + 1
	svgSizeX, svgSizeY := renderer.getSvgSize(gridCount, true)
	
	svg := NewXmlWriter(out)
	svg.
		Decl().
		Root("svg").
		Attr("width", svgSizeX).
		Attr("height", svgSizeY).
		Attr("xmlns", "http://www.w3.org/2000/svg")

	if renderer.includeBackground {
		svg.
			Elem("rect").
			Attr("width", "100%").
			Attr("height", "100%").
			Attr("fill", renderer.Property("backgroundColor")).
			Term()
	}

	gridX, gridY := 0, 0
	currentGrid := grid.Copy()
	renderer.render(svg, gridX, gridY, currentGrid, nil, "début")

	for moveIndex, move := range(moves) {
		title := fmt.Sprintf("étape %d", moveIndex + 1)

		gridX, gridY = renderer.nextGridPosition(gridX, gridY)
		renderer.render(svg, gridX, gridY, currentGrid, &move, title)
		currentGrid.Apply(move)
	}

	gridX, gridY = renderer.nextGridPosition(gridX, gridY)
	renderer.render(svg, gridX, gridY, currentGrid, nil, "fin")
	svg.Term()
}

func (renderer *SvgRenderer) render(
	svg *XmlWriter,
	gridX int,
	gridY int,
	grid *pt.Grid,
	move *pt.Move,
	title string) {

	includeTitle := title != ""
	gridCoordX, gridCoordY := renderer.gridPositionToXYCoord(gridX, gridY, includeTitle)
	
	svg.
		Elem("g").
		Attr("transform", fmt.Sprintf("translate(%g %g)", gridCoordX, gridCoordY))

	renderer.CustomRenderGrid(renderer, svg, grid, title)

	for pieceIndex := range(grid.Pieces) {
		piece := &grid.Pieces[pieceIndex]
		if move != nil && move.PieceIndex == pieceIndex {
			renderer.CustomRenderPiece(renderer, svg, piece, &move.Step)
		} else {
			renderer.CustomRenderPiece(renderer, svg, piece, nil)
		}
	}
	svg.Term()
}

File: python/svg_renderer.py
# SPDX-License-Identifier: MIT
# Copyright © 2023  Bilal Djelassi

from xml.etree import ElementTree as ET

def format_xml_attrib(value):
    if isinstance(value, str):
        return value
    if isinstance(value, int):
        return "{}".format(value)
    if isinstance(value, float):
        return "{:g}".format(value)
    raise TypeError("attribute must be str, int or float")

def create_xml_element(name, attribs={}, content="", sub_elements=[]):
    element = ET.Element(name)
    for key, value in attribs.items():
        element.set(key, format_xml_attrib(value))
    if content:
        element.text = content
    if sub_elements:
        element.extend(sub_elements)
    return element

def serialize_xml_document(document):
    return ET.tostring(document,
        xml_declaration=True,
        encoding="unicode")

class SVGRenderer:
    __slots__ = \
        'cell_gap', 'grid_gap', 'unit_size', 'title_height', \
        'offset_x', 'offset_y', 'border_radius', \
        'grids_per_row', 'grid_size_x', 'grid_size_y', \
        'include_background', 'properties'

    def __init__(self, grid):
        self.cell_gap = 4.0
        self.grid_gap = 16.0
        self.unit_size = 20.0
        self.title_height = 12.0
        self.offset_x = 0.5
        self.offset_y = 0.5
        self.border_radius = 2.0
        self.grids_per_row = 10
        self.grid_size_x = grid.size_x
        self.grid_size_y = grid.size_y
        self.include_background = False
        self.properties = {}

    @staticmethod
    def simple_theme():
        return {
            "fontFamily": "'Consolas','PT Mono','DejaVu Serif Mono',monospace",
            "fontSize": "16",
            "strokeWidthGrid": "1",
            "strokeWidthPiece": "1",
            "backgroundColor": "#ffffff",
            "textColor": "#0d1e26",
            "arrowColor": "#ffffff",
            "fillColorGrid": "#ecf4f9",
            "strokeColorGrid": "#8f9ca3",
            "fillColorHiPieceA": "#00aaff",
            "fillColorLoPieceA": "#8f9ca3",
            "fillColorHiPieceB": "#00aaff",
            "fillColorLoPieceB": "#8f9ca3",
            "fillColorHiPieceC": "#00aaff",
            "fillColorLoPieceC": "#8f9ca3",
            "fillColorHiPieceD": "#00aaff",
            "fillColorLoPieceD": "#8f9ca3",
            "strokeColorHiPieceA": "#0d1e26",
            "strokeColorLoPieceA": "#0d1e26",
            "strokeColorHiPieceB": "#0d1e26",
            "strokeColorLoPieceB": "#0d1e26",
            "strokeColorHiPieceC": "#0d1e26",
            "strokeColorLoPieceC": "#0d1e26",
            "strokeColorHiPieceD": "#0d1e26",
            "strokeColorLoPieceD": "#0d1e26",
        }

    @staticmethod
    def colorful_theme():
        return {
            "fontFamily": "'Consolas','PT Mono','DejaVu Serif Mono',monospace",
            "fontSize": "16",
            "strokeWidthGrid": "1",
            "strokeWidthPiece": "1",
            "backgroundColor": "#ffffff",
            "textColor": "#0d1e26",
            "arrowColor": "#ffffff",
            "fillColorGrid": "#ecf4f9",
            "strokeColorGrid": "#8f9ca3",
            "fillColorHiPieceA": "#ec433b",
            "fillColorLoPieceA": "#ec433b",
            "fillColorHiPieceB": "#93a707",
            "fillColorLoPieceB": "#93a707",
            "fillColorHiPieceC": "#3b98e0",
            "fillColorLoPieceC": "#3b98e0",
            "fillColorHiPieceD": "#e5b01a",
            "fillColorLoPieceD": "#e5b01a",
            "strokeColorHiPieceA": "#dc322f",
            "strokeColorLoPieceA": "#dc322f",
            "strokeColorHiPieceB": "#859900",
            "strokeColorLoPieceB": "#859900",
            "strokeColorHiPieceC": "#268bd2",
            "strokeColorLoPieceC": "#268bd2",
            "strokeColorHiPieceD": "#dba700",
            "strokeColorLoPieceD": "#dba700",
        }

    def get_grids_on_xy_axis(self, grid_count):
        if grid_count <= self.grids_per_row:
            grids_on_x_axis = grid_count
            grids_on_y_axis = 1
        else:
            grids_on_x_axis = self.grids_per_row
            grids_on_y_axis = 1 + (grid_count - 1) // self.grids_per_row
        return (grids_on_x_axis, grids_on_y_axis)

    def get_grid_size(self, include_title=False):
        grid_size_x = self.cell_gap \
             + self.grid_size_x * (self.unit_size + self.cell_gap)
        grid_size_y = self.cell_gap \
             + self.grid_size_y * (self.unit_size + self.cell_gap) \
             + self.title_height * include_title
        return (grid_size_x, grid_size_y)

    def get_svg_size(self, grid_count, include_title=False):
        grids_on_x_axis, grids_on_y_axis = self.get_grids_on_xy_axis(grid_count)
        grid_size_x, grid_size_y = self.get_grid_size(include_title)
        svg_size_x = self.grid_gap + grids_on_x_axis * (grid_size_x + self.grid_gap)
        svg_size_y = self.grid_gap + grids_on_y_axis * (grid_size_y + self.grid_gap)
        return (svg_size_x, svg_size_y)

    def grid_position_to_xy_coord(self, x, y, include_title=False):
        grid_size_x, grid_size_y = self.get_grid_size(include_title)
        grid_x = self.offset_x + self.grid_gap + x * (grid_size_x + self.grid_gap)
        grid_y = self.offset_y + self.grid_gap + self.title_height * include_title \
                + y * (grid_size_y + self.grid_gap)
        return (grid_x, grid_y)

    def piece_position_to_xy_coord(self, x, y):
        piece_x = self.cell_gap + x * (self.unit_size + self.cell_gap)
        piece_y = self.cell_gap + y * (self.unit_size + self.cell_gap)
        return (piece_x, piece_y)

    def next_grid_position(self, x, y):
        return (0, y+1) if x+1 >= self.grids_per_row else (x+1, y)

    def render_grid(self, grid):
        svg_size_x, svg_size_y = self.get_svg_size(1, False)

        svg = create_xml_element("svg", {
            "width": svg_size_x,
            "height": svg_size_y,
            "xmlns": "http://www.w3.org/2000/svg"
        })

        if self.include_background:
            svg.append(create_xml_element("rect", {
                "width": "100%",
                "height": "100%",
                "fill": self.properties["backgroundColor"],
}))

        self.render(svg, (0, 0), grid)
        return svg

    def render_grids(self, grid, moves):
        grid_count = 1 + len(moves) + 1
        svg_size_x, svg_size_y = self.get_svg_size(grid_count, include_title=True)

        svg = create_xml_element("svg", {
            "width": svg_size_x,
            "height": svg_size_y,
            "xmlns": "http://www.w3.org/2000/svg",
        })

        if self.include_background:
            svg.append(create_xml_element("rect", {
                "width": "100%",
                "height": "100%",
                "fill": self.properties["backgroundColor"],
            }))

        grid_pos_x, grid_pos_y = (0, 0)
        current_grid = grid.copy()
        self.render(svg, grid_pos_x, grid_pos_y, current_grid, None, "début")

        for index, move in enumerate(moves):
            title = "étape {}".format(index + 1)
            grid_pos_x, grid_pos_y = self.next_grid_position(grid_pos_x, grid_pos_y)
            self.render(svg, grid_pos_x, grid_pos_y, current_grid, move, title)
            current_grid.apply(move)

        grid_pos_x, grid_pos_y = self.next_grid_position(grid_pos_x, grid_pos_y)
        self.render(svg, grid_pos_x, grid_pos_y, current_grid, None, "fin")

        return svg

    def render(self, svg, grid_pos_x, grid_pos_y, grid, move, title):
        grid_coord_x, grid_coord_y = \
            self.grid_position_to_xy_coord(grid_pos_x, grid_pos_y, title is not None)

        svg_element = create_xml_element("g", {
            "transform": "translate({:g} {:g})".format(grid_coord_x, grid_coord_y),
        })
        self.custom_render_grid(svg_element, grid, title)

        for index, piece in enumerate(grid.pieces):
            if move is not None and move.piece_index == index:
                self.custom_render_piece(svg_element, piece, move.step)
            else:
                self.custom_render_piece(svg_element, piece, None)

        svg.append(svg_element)


    def custom_render_grid(self, svg_element, grid, title):
        raise NotImplemented("must be implemented in a derived class")

    def custom_render_piece(self, svg_element, piece, step):
        raise NotImplemented("must be implemented in a derived class")

